Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/common/ClusterManagerEx.java

```diff
@@ -182,18 +187,25 @@ public boolean isLIstAdditionalClusterSuccess() {
                                     "List HDInsight Cluster Error");
                         }
 
-                        return Optional.empty();
+                        return Optional.of(new ArrayList<IClusterDetail>());
                     }
                 })
                 .ifPresent(clusterDetailList -> {
                     // TODO: so far we have not a good way to judge whether it is token expired as
                     //       we have changed the way to list HDInsight clusters
                     isListClusterSuccess = clusterDetailList.size() != 0;
 
-                    clusterDetailList.addAll(hdinsightAdditionalClusterDetails);
-                    clusterDetailList.addAll(emulatorClusterDetails);
-                    ClusterMetaDataService.getInstance().addCachedClusters(clusterDetailList);
+                    List<IClusterDetail> mergedClusters = clusterDetailList.stream()
+                            // replace the duplicated cluster with linked one
+                            .map(cluster -> hdinsightAdditionalClusterDetails.stream()
+                                                .filter(linkedCluster -> linkedCluster.getName().equals(cluster.getName()))
+                                                .findFirst()
+                                                .orElse(cluster))
+                            .collect(Collectors.toList());
+
+                    mergedClusters.addAll(emulatorClusterDetails);
 
+                    ClusterMetaDataService.getInstance().addCachedClusters(mergedClusters);
```

Comment:
How about users sign in but link cluster not belonging to their subscription? It seems like we are not adding hdinsightAdditionalClusterDetails or emulatorClusterDetails if users signed in but do not have any subscription? Probably we should do a full-outer join here not left-outer join

Reply:
Good catch! Thanks DJ! And there is another issue found of not filtering the selected subscription.